Brand - Button accessibility updates

TylerJDev opened the pull request:
## Summary

Adds accessibility fixes related to [Brand-Button accessibility review](https://github.com/github/primer/issues/1682).

## List of notable changes:

- Added outline on focus when only `aria-disabled` is applied. 
- Added focus indicator and 'disabled' style for Windows High Contrast Mode (WHCM).
- Added use of 'disabled' and 'aria-disabled' separately.
- Hid SVG from screen readers. 

## What should reviewers focus on?

- `aria-disabled` and `disabled` now being separated.
- Styles for button for when `aria-disabled` is applied (focus indication, WHCM support)

## Steps to test:

<!--
Help reviewers test the feature by providing steps to reproduce the behavior.

E.g.

1. Open the # component in CI-deployed preview environment
2. Go to # story in Storybook
3. Verify that # behaves as described in the following issue.
-->

1. Open up component and use `aria-disabled` attribute control on button. Ensure element is focusable with only `aria-disabled` being applied.
2. Test the same as above, but with the "`disabled`" control, ensure that element is not focusable.
3. See if WHCM styles are applied. 
  (This can be tested within Chrome without Windows through [Chrome Devtools via the "Rendering" tab](https://developer.chrome.com/docs/devtools/rendering/emulate-css/#emulate-css-media-feature-forced-colors). Within the "Emulate CSS media feature forced-colors" section, select the "forced-colors: active" option in the select menu).

## Supporting resources (related issues, external links, etc):

- [Accessibility engineer review for Primer/Brand Button](https://github.com/github/primer/issues/1682)
- [Windows High Contrast Mode & development](https://blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors/)

## Contributor checklist:

- [ ] All new and existing CI checks pass
- [ ] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [ ] New visual snapshots have been generated / updated for any UI changes
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Screenshots:

<table>
<tr>
<th> Before</th> <th> After </th> <th> Why </th>
</tr>
<tr>
<td valign="top">

![Primer Brand button with "disabled" and "aria-disabled" applied, shows a dimmed button without a focus outline.](https://user-images.githubusercontent.com/26746305/215937632-ac98ff1e-def6-489e-860d-891b24e9e73d.png)

`disabled` and `aria-disabled` applied, no focus indication is present (due to `disabled` attribute, and existing CSS rule).
 </td>


<td valign="top">

![Primer Brand button with only "aria-disabled" applied, shows a dimmed button with a focus outline on focus.](https://user-images.githubusercontent.com/26746305/215937656-96daa9fb-e145-4979-8ac7-7ce66bc0d5a5.png)

Only `aria-disabled` is applied in this example, retains focus indicator due to element still being focusable.
</td>

<td valign="middle">

We'd want to keep the focus outline on the button only when `aria-disabled` is applied. This is because `aria-disabled` will keep the button focusable.

This keeps us in check with [WCAG 2.4.7 Focus visible](https://www.w3.org/WAI/WCAG22/Understanding/focus-Visible.html).
</td>
</tr>

<tr>
<td valign="top">

<img width="170" alt="Primer Brand button within Windows High Contrast Mode (emulated), showing a button with no focus indicator while being actively focused." src="https://user-images.githubusercontent.com/26746305/215944088-052eb1d5-d9b4-42e5-81b7-10720e5be5b0.png">

Focused button (`:focus`) within WHCM.

</td>
<td valign="top">

<img width="170" alt="Primer Brand button within Windows High Contrast Mode (emulated), showing a button with a visible focus indicator while being focused." src="https://user-images.githubusercontent.com/26746305/215944022-48cdfae4-781a-4652-b551-0aa3c1b59bc5.png">

Focused button (`:focus`), within WHCM. Focus outline is present via `outline` CSS rule.

</td>
<td valign="middle">

WHCM by default will strip `box-shadow` from elements visually, meaning that our current focus indicator that uses `box-shadow` needs to be supplemented with an `outline` style. This new rule will only be applied within WHCM (via `@media (forced-colors: active)`).
</td>
</tr>
</table>

rezrah commented on .changeset/odd-penguins-film.md:
```diff
@@ -0,0 +1,10 @@
+'@primer/react-brand': patch
+---
+
+Adds accessibility fixes.
```
Suggestion for adding some clarity on which component this update will impact:

```suggestion
Added accessibility fixes for `Button`
```

TylerJDev replied:
Good call! Should be adjusted now.

rezrah commented on packages/react/src/Button/Button.features.stories.tsx:
```diff
@@ -16,7 +16,7 @@ export const Primary = () => (
 
 export const PrimaryDisabled = () => (
-  <Button as="a" disabled variant="primary" href="#">
+  <Button as="button" disabled variant="primary" href="#">
```
Can we remove the `href` in this example too if we're using `button`?

```suggestion
  <Button as="button" disabled variant="primary">
```

TylerJDev replied:
Definitely, good call! I removed the `href` if the component was either the default (button) or `as="button"`.

rezrah commented on packages/react/src/Button/Button.module.css:
```diff
@@ -113,3 +113,17 @@
   box-shadow: none;
 }
+
+@media (forced-colors: active) {
```
TIL. Thanks for adding this, I suspect we'll probably need to do the same in a few other components 😓

TylerJDev replied:
No worries! I think it'll be a case by case basis for now. I'll have to do some research to see if there's a more global way we can apply these rules.

rezrah commented on packages/react/src/Button/Button.module.css:
```diff
@@ -113,3 +113,17 @@
+
+  /* Indicates disabled state in WHCM */
+  .Button--label.Button-label--disabled {
+    color: GrayText;
```
Question: Are our design tokens not suitable here? Or if `GrayText` is required, should we have a design token that maps to this?

TylerJDev replied:
I'm not too sure. `GrayText` here only tells WHCM that the text should be styled as if it's disabled. I wonder if we'd benefit from adding this as a token or not? As of right now, I don't think we'll be using it for anything other than color for text that should be disabled. 

Would love to know your thoughts on this.

rezrah replied:
That's good to know. I don't think this needs to be token in that case. It will create unnecessary obfuscation and also isn't something we'd want users to customise from sounds of it.

rezrah commented on packages/react/src/Button/Button.module.css:
```diff
@@ -113,3 +113,17 @@
+@media (forced-colors: active) {
+  /* Retains focus indicator when in "Windows High Contrast Mode". */
+  .Button:focus:not(.Button--disabled[disabled]) {
+    outline: 3px solid;
```
FYI, we have some tokens that could perhaps replace the `px` variant?

E.g. 

```css
--brand-borderWidth-thick
--brand-borderWidth-thicker
```

TylerJDev replied:
Ahh good to know! I replaced this with `--brand-borderWidth-thick` and made note of our tokens.

rezrah commented on packages/react/src/Button/Button.tsx:
```diff
@@ -16,6 +16,7 @@ export type ButtonBaseProps = {
   variant?: typeof ButtonVariants[number]
   size?: typeof ButtonSizes[number]
   hasArrow?: boolean
+  ariaDisabled?: boolean
```
Given that Button is a wrapper around HTMLElement, could it be confusing that we're redeclaring this rather than relying on the inherited attribute?

TylerJDev replied:
That makes sense. I've adjusted it so that we rely on `aria-disabled` instead!

rezrah reviewed (commented):
This is great, thank you @TylerJDev 🙌 

I've left some super minor feedback for you. Specifically the ones around removing the leftover `href` in the code snippets.

TylerJDev commented on packages/react/src/Button/Button.tsx:
```diff
@@ -45,44 +47,47 @@ export const Button = forwardRef(
     const [isFocused, setIsFocused] = React.useState(false)
     const Component = as || 'button'
 
+    const isDisabled =
```
Added a new variable `isDisabled`. Since `aria-disabled` can take a value of `string` or `boolean` (e.g. `aria-disabled="true"` === `aria-disabled={true}`) I made a conditional to verify that it either has a string value of `"true"` or a boolean value of `true`.

TylerJDev commented:
@rezrah - Thanks a bunch for checking out this PR! I addressed all your comments, and adjusted the PR slightly. I've included a test for `Button` too. Let me know if you have any questions, or more feedback!

rezrah reviewed (approved):
Looks great, ship it :shipit:
